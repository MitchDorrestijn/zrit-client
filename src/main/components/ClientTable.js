/**
 * React related imports
 */
import React from 'react';
import {Button} from 'reactstrap'
import {CSVLink} from 'react-csv';
import {BootstrapTable, TableHeaderColumn} from 'react-bootstrap-table';
import {renderSearchField, renderSortedColumn} from '../global/Methods';
import Tooltip from "react-simple-tooltip";

/**
 * Endpoints import
 */
import {
  getAllClients
} from '../CRUD/Client';

/**
 * Style related imports
 */
import '../../../node_modules/react-bootstrap-table/dist/react-bootstrap-table-all.min.css';
import '../css/table.css';

/**
 * Other imports
 */
import warning from '../assets/img/warning.png';


/**
 * This class takes care of rendering the client table, from here other actions can be taken like editing
 */
export default class ClientTable extends React.Component {

  /**
   * Makes a GET request to get all chauffeurs when component is mounted
   */
  componentDidMount() {
    getAllClients(this.props).then((res) => {res !== undefined && this.setState({data: res.data})});
  }

  /**
   * Setups all the data.
   * @constructor
   * @param {props} props - The given properties by the superclass
   */
  constructor(props) {
    super(props);
    this.columns = [
      {
        name: 'warningPKB',
        display: 'Warning'
      }, {
        name: 'name',
        display: 'Naam cliënt'
      }, {
        name: 'pkb',
        display: 'PKB cliënt'
      }, {
        name: 'totalMeters',
        display: 'Gemaakte meters'
      }, {
        name: 'priceToPay',
        display: 'Te betalen door cliënt'
      }
    ];
    this.state = {
      sortColumnName: undefined,
      sortOrder: undefined,
      disableButtons: true,
      data: []
    };
  }

  /**
   * Renders the buttons above the table
   */
  renderButtons = () => {
    return (<div>
      <Button color="primary" className='crud-btn'>Toevoegen</Button>
      <Button disabled={this.state.disableButtons} color="primary" className='crud-btn'>Bewerken</Button>
    </div>);
  };

  /**
   * Renders the warning icon based on the warning state
   */
  warningFormatter = (cell) => {
    switch (cell) {
      case true:
        return (
          <div className="tooltip_wrapper">
            <Tooltip placement="right" content="PKB bijna op!">
              <img className="warning_icon" src={warning} alt="warning" />
            </Tooltip>
          </div>
        )
      default:
        return (
          <span>-</span>
        )
    }
  }

  /**
   * Renders the view for the user
   */
  render() {
    const tableOptions = {
      sortColumnName: this.state.sortName,
      sortOrder: this.state.sortOrder,
      renderSortedColumn: renderSortedColumn,
      noDataText: 'Geen resultaten gevonden',
      searchField: renderSearchField,
      btnGroup: this.renderButtons
    };

    return (<div>
      <BootstrapTable className="clientTable" search={true} data={this.state.data} options={tableOptions} selectRow={{
          mode: 'radio',
          onSelect: this.onSelect
        }} ref='table'>

        <TableHeaderColumn hidden={true} dataField="clientId" isKey={true}>
          ID
        </TableHeaderColumn>

        <TableHeaderColumn dataFormat={this.warningFormatter} width="70" dataField="warningPKB" dataSort={true}>
          Warning
        </TableHeaderColumn>

        <TableHeaderColumn width="200" dataField="name" dataSort={true}>
          Naam
        </TableHeaderColumn>

        <TableHeaderColumn width="75" dataField="pkb" dataSort={true}>
          PKB
          <br/>
          cliënt &#x2195;
        </TableHeaderColumn>

        <TableHeaderColumn width="75" dataField="totalMeters" dataSort={true}>
          Gemaakte
          <br/>
          km's' &#x2195;
        </TableHeaderColumn>

        <TableHeaderColumn width="130" dataField="priceToPay" dataSort={true}>
          Te betalen
          <br/>
          door cliënt &#x2195;
        </TableHeaderColumn>

      </BootstrapTable>

      <CSVLink data={this.state.data} filename={"clienten_overview"} className="btn btn-primary crud-btn" target="">
        Export als CSV
      </CSVLink>
    </div>);
  };
}
