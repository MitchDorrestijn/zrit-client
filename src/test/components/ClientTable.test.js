import React from 'react';
import {shallow} from 'enzyme';
import toJSON from 'enzyme-to-json';
import ClientTable from '../../main/components/ClientTable';

const wrapper = shallow(<ClientTable history="" routes="" />);

const push = jest.fn();
wrapper.setProps({ history: { push } });

test('CRUDTable component should match snapshot', () => {
  expect(toJSON(wrapper)).toMatchSnapshot();
});

test('Should render seven TableHeaderColumn components', () => {
  expect(wrapper.find('TableHeaderColumn').length).toBe(6);
});

test('Should render one BootstrapTable component', () => {
  expect(wrapper.find('BootstrapTable').length).toBe(1);
});

test('Should render one CSVLink component', () => {
  expect(wrapper.find('CSVLink').length).toBe(1);
});
