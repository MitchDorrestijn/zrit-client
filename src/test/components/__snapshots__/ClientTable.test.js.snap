// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CRUDTable component should match snapshot 1`] = `
<div>
  <BootstrapTable
    autoCollapse={
      Object {
        "filter": false,
        "search": false,
        "sort": false,
      }
    }
    bodyContainerClass={null}
    bordered={true}
    cellEdit={
      Object {
        "afterSaveCell": undefined,
        "beforeSaveCell": undefined,
        "blurToEscape": false,
        "blurToSave": false,
        "mode": "none",
        "nonEditableRows": undefined,
      }
    }
    columnFilter={false}
    condensed={false}
    containerClass={null}
    csvFileName="spreadsheet.csv"
    data={Array []}
    deleteRow={false}
    expandColumnOptions={
      Object {
        "expandColumnBeforeSelectColumn": true,
        "expandColumnComponent": undefined,
        "expandColumnVisible": false,
        "expandedColumnHeaderComponent": undefined,
      }
    }
    exportCSV={false}
    fetchInfo={
      Object {
        "dataTotalSize": 0,
      }
    }
    footer={false}
    headerContainerClass={null}
    height="100%"
    hover={false}
    ignoreSinglePage={false}
    insertRow={false}
    keyBoardNav={false}
    multiColumnSearch={false}
    multiColumnSort={1}
    options={
      Object {
        "btnGroup": [Function],
        "noDataText": "Geen resultaten gevonden",
        "renderSortedColumn": [Function],
        "searchField": [Function],
        "sortColumnName": undefined,
        "sortOrder": undefined,
      }
    }
    pagination={false}
    printable={false}
    renderAlert={true}
    replace={false}
    search={true}
    selectRow={
      Object {
        "mode": "radio",
        "onSelect": undefined,
      }
    }
    striped={false}
    tableBodyClass={null}
    tableContainerClass={null}
    tableFooterClass={null}
    tableHeaderClass={null}
    trClassName=""
    version="3"
    withoutTabIndex={false}
  >
    <TableHeaderColumn
      className=""
      columnClassName=""
      columnTitle={false}
      csvFieldType="string"
      dataAlign="left"
      dataField="clientId"
      dataSort={false}
      defaultASC={false}
      editColumnClassName=""
      editable={true}
      expandable={true}
      filterFormatted={false}
      headerTitle={true}
      hidden={true}
      hiddenOnInsert={false}
      invalidEditColumnClassName=""
      isKey={true}
      keyValidator={false}
      searchable={true}
      sortIndicator={true}
      width={null}
    >
      ID
    </TableHeaderColumn>
    <TableHeaderColumn
      className=""
      columnClassName=""
      columnTitle={false}
      csvFieldType="string"
      dataAlign="left"
      dataField="warningPKB"
      dataSort={true}
      defaultASC={false}
      editColumnClassName=""
      editable={true}
      expandable={true}
      filterFormatted={false}
      headerTitle={true}
      hidden={false}
      hiddenOnInsert={false}
      invalidEditColumnClassName=""
      isKey={false}
      keyValidator={false}
      searchable={true}
      sortIndicator={true}
      width="50"
    >
      Warning
    </TableHeaderColumn>
    <TableHeaderColumn
      className=""
      columnClassName=""
      columnTitle={false}
      csvFieldType="string"
      dataAlign="left"
      dataField="name"
      dataSort={true}
      defaultASC={false}
      editColumnClassName=""
      editable={true}
      expandable={true}
      filterFormatted={false}
      headerTitle={true}
      hidden={false}
      hiddenOnInsert={false}
      invalidEditColumnClassName=""
      isKey={false}
      keyValidator={false}
      searchable={true}
      sortIndicator={true}
      width="100"
    >
      Naam
    </TableHeaderColumn>
    <TableHeaderColumn
      className=""
      columnClassName=""
      columnTitle={false}
      csvFieldType="string"
      dataAlign="left"
      dataField="pkb"
      dataSort={true}
      defaultASC={false}
      editColumnClassName=""
      editable={true}
      expandable={true}
      filterFormatted={false}
      headerTitle={true}
      hidden={false}
      hiddenOnInsert={false}
      invalidEditColumnClassName=""
      isKey={false}
      keyValidator={false}
      searchable={true}
      sortIndicator={true}
      width="130"
    >
      PKB
      <br />
      cliënt ↕
    </TableHeaderColumn>
    <TableHeaderColumn
      className=""
      columnClassName=""
      columnTitle={false}
      csvFieldType="string"
      dataAlign="left"
      dataField="totalMeters"
      dataSort={true}
      defaultASC={false}
      editColumnClassName=""
      editable={true}
      expandable={true}
      filterFormatted={false}
      headerTitle={true}
      hidden={false}
      hiddenOnInsert={false}
      invalidEditColumnClassName=""
      isKey={false}
      keyValidator={false}
      searchable={true}
      sortIndicator={true}
      width="130"
    >
      Gemaakte
      <br />
      km's' ↕
    </TableHeaderColumn>
    <TableHeaderColumn
      className=""
      columnClassName=""
      columnTitle={false}
      csvFieldType="string"
      dataAlign="left"
      dataField="priceToPay"
      dataSort={true}
      defaultASC={false}
      editColumnClassName=""
      editable={true}
      expandable={true}
      filterFormatted={false}
      headerTitle={true}
      hidden={false}
      hiddenOnInsert={false}
      invalidEditColumnClassName=""
      isKey={false}
      keyValidator={false}
      searchable={true}
      sortIndicator={true}
      width="130"
    >
      Te betalen
      <br />
      door cliënt ↕
    </TableHeaderColumn>
  </BootstrapTable>
  <CSVLink
    className="btn btn-primary crud-btn"
    data={Array []}
    filename="clienten_overview"
    separator=","
    target=""
    uFEFF={true}
  >
    Export als CSV
  </CSVLink>
</div>
`;
